File: src/09.tree/25.653.ts
import { TreeNode } from "./01.100";

export function findTarget1(root: TreeNode | null, k: number): boolean {
    // 利用 hashMap
    const m = new Map<number, boolean>();

    const dfs = (node: TreeNode | null) => {
        if (!node) {
            return false;
        }

        if (m.has(k - node.val)) {
            return true;
        }

        m.set(node.val, true);

        return dfs(node.left) || dfs(node.right);
    };

    return dfs(root);
}

export function findTarget(root: TreeNode | null, k: number): boolean {
    const list: number[] = [];

    const bts = (node: TreeNode | null) => {
        if (!node) {
            return;
        }

        bts(node.left);
        list.push(node.val);
        bts(node.right);
    };

    bts(root);

    // 应该 sort 下 list
    let l = 0;
    let r = list.length - 1;

    while (l < r) {
        const sum = list[l] + list[r];
        if (sum === k) {
            return true;
        } else if (sum > k) {
            r--;
        } else {
            l++;
        }
    }

    return false;
}

File: src/09.tree/25.653.go
package tree

func FindTarget1(root *TreeNode, k int) bool {
	var dfs func(node *TreeNode) bool
	m := map[int]bool{}

	dfs = func(node *TreeNode) bool {
		if node == nil {
			return false
		}

		if _, ok := m[k-node.Val]; ok {
			return true
		}

		m[node.Val] = true

		return dfs(node.Left) || dfs(node.Right)
	}

	return dfs(root)
}

func FindTarget(root *TreeNode, k int) bool {
	var bts func(node *TreeNode)
	list := []int{}

	bts = func(node *TreeNode) {
		if node == nil {
			return
		}
		bts(node.Left)
		list = append(list, node.Val)
		bts(node.Right)
	}

	bts(root)

	l, r := 0, len(list)-1

	for l < r {
		sum := list[l] + list[r]
		if sum == k {
			return true
		} else if sum > k {
			r--
		} else {
			l++
		}
	}

	return false
}
